cmd/ndm_daemonset/probe/eventhandler.go

```diff
@@ -72,32 +99,48 @@ func (pe *ProbeEvent) addBlockDeviceEvent(msg controller.EventMessage) {
 
 // deleteBlockDeviceEvent deactivate blockdevice resource using uuid from etcd
 func (pe *ProbeEvent) deleteBlockDeviceEvent(msg controller.EventMessage) {
-	blockDeviceOk := pe.deleteBlockDevice(msg)
-
-	// when one disk is removed from node and entry related to
-	// that disk is not present in etcd,  in that case it
-	// again rescan full system and update etcd accordingly.
-	if !blockDeviceOk {
-		go pe.initOrErrorEvent()
-	}
-}
-
-func (pe *ProbeEvent) deleteBlockDevice(msg controller.EventMessage) bool {
-	bdList, err := pe.Controller.ListBlockDeviceResource()
+	bdAPIList, err := pe.Controller.ListBlockDeviceResource(false)
 	if err != nil {
 		klog.Error(err)
-		return false
 	}
-	ok := true
-	for _, diskDetails := range msg.Devices {
-		existingBlockDeviceResource := pe.Controller.GetExistingBlockDeviceResource(bdList, diskDetails.UUID)
+
+	isDeactivated := true
+	isGPTBasedUUIDEnabled := pe.Controller.FeatureGates.IsEnabled(features.GPTBasedUUID)
+
+	for _, device := range msg.Devices {
+		// create the new UUID for removing the device
+		if isGPTBasedUUIDEnabled {
+			_, ok := pe.Controller.BDHierarchy[device.DevPath]
+			if !ok {
+				klog.Infof("Disk %s not in hierarchy", device.DevPath)
+				// not in hierarchy continue
+				continue
+			}
+			// remove from the hierarchy
+			delete(pe.Controller.BDHierarchy, device.DevPath)
+
+			uuid, ok := generateUUID(*device)
+			if !ok {
+				klog.Info("could not recreate UUID while removing device")
+				continue
+			}
+			// use the new UUID for deactivating the blockdevice
+			device.UUID = uuid
+		}
+
+		existingBlockDeviceResource := pe.Controller.GetExistingBlockDeviceResource(bdAPIList, device.UUID)
```

Comment:
Will this query fail for the claimed disks, new uuid generation is not done?  If a device is used by cstor, it might still get a new uuid because of the filesystem or the partitions.

Reply:
A claimed device will always have UUID.

> If a device is used by cstor, it might still get a new uuid because of the filesystem or the partitions.

Can this case happen in fresh installations.?

NOTE: Device already used by cstor are not handled in this PR.